src/main/scala/com/fulcrumgenomics/bam/AnnotateMolecularBarcodes.scala

```diff
@@ -0,0 +1,148 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.cmdline.{ClpGroups, FgBioTool}
+import com.fulcrumgenomics.util.ReadSegment.{MolecularBarcode, ReadSegment, Template}
+import com.fulcrumgenomics.util.ReadStructure
+import dagr.commons.CommonsDef.PathToBam
+import dagr.commons.io.Io
+import dagr.commons.util.LazyLogging
+import dagr.sopt._
+import dagr.sopt.cmdline.ValidationException
+import htsjdk.samtools._
+import htsjdk.samtools.util._
+
+import scala.collection.JavaConversions._
+
+@clp(description =
```

Comment:
I think we need to do two things:
1) Give this a better name
2) Make it clear how this is different from `AnnotateBamWithUmis`

On the naming front I would suggest:
a) We use `UMI` consistently as our term, instead of Molecular Barcode, Molecular Index, etc. so that it's clear we're talking about the same thing
b) Using a word like `Extract` or a synonym to indicate that the UMIs are being extracted from the reads in the BAM, not an external source

Perhaps calling it `ExtractUmisFromReads` or similar would be better?

On 2, I think it would be good to make the distinction in the first line of the documentation in both tools.  For this one perhaps we write `Extracts unique molecular indices from reads in a BAM file into tags.`, and edit `AnnotateBamWithUmis` to read `Annotates existing BAM files with UMIs (Unique Molecular Indices, aka Molecular IDs, Molecular barcodes) from a separate FASTQ file.`